Bai tap 1 Tensor co ban (Lam ngay 15/9/26)


In [ ]:
import torch

a = torch.tensor(
    [[1, 2, 3],
    [4, 5, 6]],
dtype=torch.int32
)

print("Kich thuoc:", a.shape)
print("So chieu:", a.ndim)
print("Kieu du lieu:", a.dtype)


Kich thuoc: torch.Size([2, 3])
So chieu: 2
Kieu du lieu: torch.int32


Bài 2 Tạo các Tensor thuong dung

Sự khác nhau giữa torch.rand() và torch.randn()
Lần 1 thử:
*   tensor([[0.2554, 0.9457, 0.1978, 0.3262],
        [0.5557, 0.0949, 0.2758, 0.5318],
        [0.9520, 0.4856, 0.5941, 0.9326]])
* tensor([[-1.4637, -0.6646,  0.2391,  0.6420],
        [-0.3704,  0.6880, -1.9906, -0.5062],
        [-0.3922,  0.2462, -1.1855,  2.3609]])

Lần 2 thử:
*   tensor([[0.0807, 0.6790, 0.8132, 0.6067],
        [0.3472, 0.1089, 0.9079, 0.0321],
        [0.4941, 0.1909, 0.4478, 0.2232]])
tensor([[-0.3439, -0.0919, -0.6749, -1.0449],
        [ 1.0933, -0.1935,  0.3717,  0.2723],
        [-1.2833, -0.1863,  3.2850,  0.3702]])

=> Rand sẽ sinh ra số ngẫu nhiên >0 còn => Randn sẽ có thêm số âm lẫn số dương






In [ ]:
import torch
a = torch.rand(3, 4)
b = torch.randn(3, 4)
c = torch.zeros(2, 3)
d = torch.ones(2, 3)
e = torch.eye(4)
print(a)
print(b)
print(c)
print(d)
print(e)

#Một Tensor 5x5 chứa toàn giá trị 10
tensor_10 = torch.full((5, 5), 10)
print(tensor_10)

tensor([[0.1617, 0.4881, 0.8660, 0.1154],
        [0.2327, 0.8210, 0.6484, 0.0017],
        [0.6673, 0.9363, 0.5632, 0.4093]])
tensor([[-0.4202, -0.1531,  0.4367, -0.4305],
        [-0.4849,  0.4637,  0.9733,  0.6607],
        [-0.6257,  1.2296, -0.0066,  1.4532]])
tensor([[0., 0., 0.],
        [0., 0., 0.]])
tensor([[1., 1., 1.],
        [1., 1., 1.]])
tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])
tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


Bài 3: Kiểm tra thông tin Tensor

torch.size([2,3,4]) có nghĩa là 2 ma trận trong đó mỗi ma trận có 3 hàng và 4 cột. và phương thức Size() có nghĩa là cung cấp kích thước trên từng chiều. Giúp biết được cấu trúc tổ chức dữ liệu bên trong Tensor đó như thế nào trước khi thực hiện các pháp toán

In [ ]:
import torch
a = torch.zeros(2, 3, 4)
print("Tensor:")
print(a)
print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)

Tensor:
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions: 3
Data type: torch.float32


Bai 4: Slicing va Indexing 

a) với a[1] thực hiện chỉ mục tại chiều đầu tiên nghĩa là chọn ra ma trận thứ 2 với index = 1 nên kết quả thu được là ma trận Thứ 2 trong tổng số 3 ma trận và có 4 hàng và 5 cột
- b) với a[1:, 2:4] thì đây là phép cắt lát trên từng chiều
- Với chiều đầu tiên (chiều 0) là (1:) thì lấy từ chỉ số 1 đến hết nghĩa là lấy ma trận thứ 2,3 bỏ ma trận 1 là chỉ số 0
- Chiều 1 (2:4) là chỉ lấy hàng 2 và trước 4 (gồm hàng 2 và 3) 
- Chiều 2 (kh viết): Mặc định giữ nguyên toàn bộ là đầy đủ 5 cột


In [10]:
import torch
a = torch.randn(3, 4, 5)
print("Original tensor:")
print(a)
print("\nTensor a[1]:")
print(a[1])
print("\nTensor a[1:, 2:4]:")
print(a[1:, 2:4])
print("Hai phan tu cuoi tren chieu thu ba:")
print(a[:, :, -2:])

Original tensor:
tensor([[[-0.5560, -0.4237, -2.5960,  0.4615, -0.4851],
         [ 0.5651,  1.4733, -0.5839,  0.4033,  1.0384],
         [ 0.9134,  1.3202, -0.5204, -0.4657,  1.7597],
         [-1.8703,  1.5767, -1.2567,  0.2981,  0.7613]],

        [[-1.4070,  0.7032, -0.8829, -1.3635, -1.2456],
         [-0.5289, -0.8735,  0.4224, -2.0813,  1.5878],
         [-0.0899,  1.4312, -1.1279,  1.3534, -0.2390],
         [-0.9096, -1.5251,  0.6879, -1.1297,  0.8646]],

        [[ 2.1171,  0.4117,  0.1767, -1.9107, -0.1561],
         [-0.9767,  0.1803, -1.3415, -0.5797,  0.5739],
         [-0.1862, -1.0275,  0.7413, -0.7539, -0.2815],
         [ 0.8035, -0.2748, -0.8833,  0.5518,  0.2618]]])

Tensor a[1]:
tensor([[-1.4070,  0.7032, -0.8829, -1.3635, -1.2456],
        [-0.5289, -0.8735,  0.4224, -2.0813,  1.5878],
        [-0.0899,  1.4312, -1.1279,  1.3534, -0.2390],
        [-0.9096, -1.5251,  0.6879, -1.1297,  0.8646]])

Tensor a[1:, 2:4]:
tensor([[[-0.0899,  1.4312, -1.1279,  1.3534, -0.2

Bài 5: Resphae và Flatten

In [13]:
from torch._inductor import cudagraph_trees
import torch
a = torch.randn(3, 4)
print("Original:")
print(a)
print("\nFlatten:")
print(a.ravel())
a_reshaped = a.reshape(3, 2, 2)
print("\nReshape to 3 x 2 x 2:")
print(a_reshaped)

#Kiem tra so phan tu
print("So phan tu truoc khi reshape:", a.numel())
print("So phan tu sau khi reshape:", a_reshaped.numel())
#2. Thu reshape Tensor thanh kich thuoc 2x6
a_2x6 = a.reshape(2, 6)
print("Reshape thanh 2 x6:")
print(a_2x6)

Original:
tensor([[ 0.6532,  1.0446,  2.1661,  1.6372],
        [ 1.2030,  0.0778, -0.4809,  0.2246],
        [-0.6636, -0.8668, -0.0956,  0.6740]])

Flatten:
tensor([ 0.6532,  1.0446,  2.1661,  1.6372,  1.2030,  0.0778, -0.4809,  0.2246,
        -0.6636, -0.8668, -0.0956,  0.6740])

Reshape to 3 x 2 x 2:
tensor([[[ 0.6532,  1.0446],
         [ 2.1661,  1.6372]],

        [[ 1.2030,  0.0778],
         [-0.4809,  0.2246]],

        [[-0.6636, -0.8668],
         [-0.0956,  0.6740]]])
So phan tu truoc khi reshape: 12
So phan tu sau khi reshape: 12
Reshape thanh 2 x6:
tensor([[ 0.6532,  1.0446,  2.1661,  1.6372,  1.2030,  0.0778],
        [-0.4809,  0.2246, -0.6636, -0.8668, -0.0956,  0.6740]])


Bai 6: Phep toán và nhân ma trận
 - a - Phân biệt phép toán a*b và a@b.T
    - Phép a * b là phép nhân từng phần tử, nó sẽ láy phần tử ở hàng i, cột j của a nhân trực tiếp với phần tử ở hàng i cột j tương ứng với b và kích thước sau khi nhân vẫn giữ nguyên 2 x 3
    - Phép a@b.T là nhân ma trận đại số tuyến tính, nó sẽ lấy từng hàng của ma trận a nhân với từng cột của ma trận b.T và .T ở đây là bắt buộc có để đảo chiều ma trận b từ (2x3) thành (3x2) thì mới đủ điều kiện để nhân sau khi nhân thì kích thước mới là 2 x 2

- b. Cho biết shape của kết quả a@b.T.
    - Kết quả của shape a@ b.T. là [2,2]

In [18]:
import torch
a = torch.randn(2, 3)
b = torch.randn(2, 3)
print("a:")
print(a)
print("b:")
print(b)
print("\na + b:")
print(a + b)
print("\na / b:")
print(a / b)
print("\na^2:")
print(a ** 2)
print("\nNhan ma tran: ")
print(a*b)
print("\nMatrix multiplication:")
print(a @ b.T)

c = a @ b.T
print("Kich thuoc shape cua c:", c.shape)
torch.Size([2, 2])

a:
tensor([[-0.6930, -1.3561,  0.2258],
        [-1.6013, -0.4580,  0.8972]])
b:
tensor([[-1.1273,  0.1875, -0.0305],
        [-0.2834, -2.2694, -0.4959]])

a + b:
tensor([[-1.8203, -1.1686,  0.1953],
        [-1.8848, -2.7274,  0.4013]])

a / b:
tensor([[ 0.6148, -7.2319, -7.3969],
        [ 5.6494,  0.2018, -1.8093]])

a^2:
tensor([[0.4803, 1.8390, 0.0510],
        [2.5642, 0.2098, 0.8049]])

Nhan ma tran: 
tensor([[ 0.7812, -0.2543, -0.0069],
        [ 0.4539,  1.0395, -0.4449]])

Matrix multiplication:
tensor([[0.5200, 3.1620],
        [1.6918, 1.0485]])
Kich thuoc shape cua c: torch.Size([2, 2])


torch.Size([2, 2])

Bài 7: Các hàm thống kê trên Tensor
- dim = 0
    - Tính theo chiều dọc theo các cột, đi qua các hàng và gồm 4 giá trị đại diện cho cả 4 cột và sizee lúc này sẽ được cộng dồn từ hàng trên xuống hàng dưới
    - Bên cạnh đó do tensor có 4 cột nên kết quả trả về là một vector gồm 4 phần tử
- dim = 1
    - Tính theo chiều ngang theo các hàng, đi qua các cột và gồm 3 giá trị đại diện cho 3 hàng và size sẽ được cộng dồn từ cột trái sang cột phải
    - Bên cạnh đó do có 3 hàng, kết quả trả về một vector gồm 3 phần tử

In [23]:
import torch
a = torch.randn(3, 4)
print(a)
print("\nMean:")
print(torch.mean(a, dim=0))
print("\nStandard deviation:")
print(torch.std(a, dim=0))
print("\nCumulative sum:")
print(torch.cumsum(a, dim=0))

print("\nMean dim = 1:")
print(torch.mean(a, dim=1))
print("\nStandard deviation dim = 1:")
print(torch.std(a, dim=1))
print("\nCumulative sum dim = 1:")
print(torch.cumsum(a, dim=1))

tensor([[ 0.9645, -1.0055,  1.5024, -0.8609],
        [ 0.6543, -0.0492, -0.4889, -0.9717],
        [ 0.8730,  1.0775,  0.5949, -0.5479]])

Mean:
tensor([ 0.8306,  0.0076,  0.5361, -0.7935])

Standard deviation:
tensor([0.1593, 1.0427, 0.9970, 0.2198])

Cumulative sum:
tensor([[ 0.9645, -1.0055,  1.5024, -0.8609],
        [ 1.6188, -1.0547,  1.0135, -1.8326],
        [ 2.4918,  0.0228,  1.6083, -2.3805]])

Mean dim = 1:
tensor([ 0.1501, -0.2139,  0.4994])

Standard deviation dim = 1:
tensor([1.2714, 0.6906, 0.7257])

Cumulative sum dim = 1:
tensor([[ 0.9645, -0.0410,  1.4614,  0.6005],
        [ 0.6543,  0.6051,  0.1162, -0.8555],
        [ 0.8730,  1.9505,  2.5454,  1.9974]])


Bài 8: Autograd cơ bản
- Tính thủ công đạo hàm của y = x2 tại x = 3.6 và so sánh với x.grad.
    - Đạo hàm của y =x^2 => 2x và tại x = 3.6 nên y = 7.2
    - Với x.grad cũng = 7.2 nên ta thấy 2 kết quả giống nhau
- Đổi hàm thành y = 3*x**2 + 2*x + 1 và kiểm tra gradient.
    - Giá trị gradient được tính ra x.grad = 23.6 và tính thủ công cũng = với tính tay

In [26]:
import torch
x = torch.tensor(
3.6,
requires_grad=True
)
y = 3*x**2 + 2*x + 1
y.backward()
print("x =", x)
print("y =", y)
print("x.grad =", x.grad)

x = tensor(3.6000, requires_grad=True)
y = tensor(47.0800, grad_fn=<AddBackward0>)
x.grad = tensor(23.6000)


Bài 9. Gradient Descent tìm hệ số đa thức
- a. So sánh ba hệ số học được với [1,2,3]
    - Trước khi huấn luyện: Trọng số w được khởi tạo ngẫu nhiên nên có giá trị sai lệch lớn so với bộ hệ số [1,2,3] ban đầu
    - Sau 1000 epochs huấn luyện: Hệ số của x^2 hội tụ nhanh và sát nhất (đạt 0.9363 gần với 1) vì bậc 2 có độ biến thiên rớt nhất nên ảnh hưởng mạnh nhất đến hàm mất mát MSE
        - Hệ số x và hệ số tự do cũng đều tự điều chỉnh và tiến về gần về phía mục tiêu (lần lượt tiệm cận về mốc 2 và 3)
    - Từ đó ta có thể thấy Thuật toán tối ưu Gradient Descent đã hoạt động hiệu quả giúp mô hình tự học và tìm lại gần đúng hàm đa thức gốc ban đầu từ dữ liệu mẫu. Nếu tăng lần lặp (epoch) lên 2000 - 3000 thì bộ hệ số này sẽ hội tụ chính xác tuyệt đối về [1.0, 2.0, 3.0]

- b. Giải thích vai trò của optimizer.zero_grad(), mse.backward() và optimizer.step().
    - optimizer.zero_grad(): có vai trò đặt lại gradient của tất cả các tham số cần học về giá trị 0 trước khi bắt đầu tính toán cho vòng lặp mới. Nó bắt buộc phải có vì cơ chế mặc định là cộng dồn gradient qua mỗi lần gọi .backward(). Nếu không xóa về 0 ở đầu mỗi epoch, gradient của vòng lặp hiện tại sẽ bị cộng dè lên gradient của các vòng lặp trước đó nên sẽ khiến các bước nhảy bị sai lệch hoàn toàn và mô hình không thể hội tụ. Kết quả tính được sẽ được lưu trữ tự động vào thuộc tính w.grad.
    - mse.backward(): là lan truyền ngược có vai trò thực hiện quá trình lan truyền ngược để tính toán cho hàm (gradient) của hàm mất mát MSE theo từng hệ số trong w,
    - optimizer.step(): là Cập nhật trọng số và có vai trò thực hiện thay đổi, cập nhật giá trị của các hệ số w nhằm giảm thiểu sai số. Nên nhờ lệnh này mà sau mỗi epoch, các giá trị trong w ngày càng nhích gần hơn về bộ số [1,2,3].

In [44]:
import numpy as np
import torch
polynomial = np.poly1d([1, 2, 3])
N = 20

X = np.random.randn(N, 1) * 5
Y = polynomial(X)

XX = np.hstack([
    X * X,
    X,
    np.ones_like(X)
])

w = torch.randn(3, 1, requires_grad=True)
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)

print("Initial coefficients:")
print(w)

for epoch in range(1000):
    optimizer.zero_grad()
    y_pred = x @ w
    mse = torch.mean(torch.square(y - y_pred))
    mse.backward()
    optimizer.step()
print("Learned coefficients:")
print(w)

Initial coefficients:
tensor([[-0.6690],
        [-0.5946],
        [-0.3786]], requires_grad=True)
Learned coefficients:
tensor([[1.0130],
        [1.9376],
        [1.5805]], requires_grad=True)


Bài 10: Giải hệ phương trình bằng Autograd
a) Kiểm tra nghiệm thu được với bốn phương trình: A+B=9, C-D=1, A+C=8, B-D=2
   - A gần bằng 4.4434
    - B gần bằng 4.5565
    - C gần bằng 3.5566
    - D gần bằng 2.5565
- Vậy khi thay các giá trị trên vào 4 phương trình theo đề cho đều thỏa mãn

b) Nhận xét vì sao bài toán có nhiều nghiệm:
- Ta thấy có 3 phương trình độc lập tuyến tính nhưng có tới 4 ẩn số là A B C D nên ta có số ẩn nhiều hơn số phương trình (4 > 3) nên có 1 bậc tự do và sẽ tôn tại vô số nghiệm
- Ngoài ra do bài toán có vô số nghiệm, đáy của hàm mất mát không phải là một điểm đơn lẻ mà là cả một đường thẳng liên tục trong không gian 4 chiều. Các biến A B C D ban đầu được gán vào giá trị ngẫu nhiên. Tùy thuộc vào thời điểm khởi tạo ban đầu, thuật toán Gradient Des sẽ trượt xuống và hội tụ tạo một điểm bất kỳ trên đường thẳng của nghiệm đó. Vì vậy mỗi lần chạy lại chương trình thì ta lại nhận một bộ nghiệm số khác nhau nhưng tất cả đều đúng



In [10]:
from numpy._typing import _nested_sequence
import random
import torch
A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)
EPOCHS = 2000
optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)
for _ in range(EPOCHS):
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2
    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4
    optimizer.zero_grad()
    sqerr.backward()
    optimizer.step()
print("A =", A)
print("B =", B)
print("C =", C)
print("D =", D)

print("Nghiem thu duoc voi bon phuong trinh:")
print("PT1: (A + B = 9):", A.item() + B.item())
print("PT 2 (C - D = 1):", C.item() - D.item())
print("PT 3 (A + C = 8):", A.item() + C.item())
print("PT 4 (B - D = 2):", B.item() - D.item())
print("Tong sai so:", sqerr.item())


A = tensor(4.4299, requires_grad=True)
B = tensor(4.5701, requires_grad=True)
C = tensor(3.5701, requires_grad=True)
D = tensor(2.5701, requires_grad=True)
Nghiem thu duoc voi bon phuong trinh:
PT1: (A + B = 9): 8.999939918518066
PT 2 (C - D = 1): 1.000014066696167
PT 3 (A + C = 8): 7.99996018409729
PT 4 (B - D = 2): 1.9999938011169434
Tong sai so: 5.4504312174685765e-09


Bài 11: Viết chương trình PyTorch để thực hiện các yêu cầu:

In [ ]:
import numpy as np  
import torch    

print("YC 1: Tạo tensor X có 24 phần tử và reshape lần lượt thành 4 x6, 2 x 12 và 2 x3 x4")

x = torch.arange(24)
print("Tensor x ban dau: ", x)
print("Shape ban dau: ", x.shape)

x_4x6 = x.reshape(4, 6)
print("Reshape 4 x 6: \n", x_4x6)

x_2x12 = x.reshape(2, 12)
print("Reshape 2 x 12: \n", x_2x12)

x_2x3x4 = x.reshape(2, 3, 4)
print("Reshape 2 x 3 x 4: \n", x_2x3x4)

print("YC2: CÁC PHÉP TOÁN TRÊN TENSOR A VÀ B (3 x 4)")
torch.manual_seed(42)
A = torch.randint(1, 10, (3, 4)).float()
B = torch.randint(1, 10, (3, 4)).float()

print("Tensor A:\n", A)
print("Tensor B:\n", B)
print("\nPhép cộng A + B:\n", A + B)
print("\nNhân từng phần tử A * B:\n", A * B)
print("\nChia từng phần tử A / B:\n", A / B)
# Trung bình theo cột
mean_col_A = torch.mean(A, dim=0)
print("\n4. Trung bình theo cột của A (dim=0):\n", mean_col_A)
# Độ lệch chuẩn theo cột
std_col_A = torch.std(A, dim=0)
print("\n5. Độ lệch chuẩn theo cột của A (dim=0):\n", std_col_A)

print("YC3: TÍNH ĐẠO HÀM y = 2x^2 + 5x + 3 TẠI x = 2")
x = torch.tensor(2.0, requires_grad=True)
y = 2 * x**2 + 5 * x + 3
y.backward()
print(f"Giá trị tính bằng Autograd (x.grad): {x.grad.item()}")

# Tính thủ công:
# y' = d(2x^2 + 5x + 3)/dx = 4x + 5
# Tại x = 2: y'(2) = 4*(2) + 5 = 13
manual_grad = 4 * 2 + 5
print(f"Giá trị tính thủ công: 4*2 + 5 = {manual_grad}")
print(
    f"So sánh: {'Trùng khớp hoàn toàn' if x.grad.item() == manual_grad else 'Khác nhau'}"
)

print("YC 4: TÌM HỆ SỐ ĐA THỨC y = 2x^2 - 3x + 5")
np.random.seed(42)
torch.manual_seed(42)
polynomial = np.poly1d([2, -3, 5]) #đa thức theo đề
N = 20

X_data = np.random.randn(N, 1) * 5
Y_data = polynomial(X_data)
# Ma trận đặc trưng [X^2, X, 1]
XX = np.hstack([X_data * X_data, X_data, np.ones_like(X_data)])

# Khởi tạo trọng số ngẫu nhiên w = [w1, w2, w3]
w = torch.randn(3, 1, requires_grad=True)
x_tensor = torch.tensor(XX, dtype=torch.float32)
y_tensor = torch.tensor(Y_data, dtype=torch.float32)
print("Hệ số ngẫu nhiên ban đầu w:\n", w.detach().numpy().flatten())
# Tối ưu hóa bằng NAdam
optimizer = torch.optim.NAdam([w], lr=0.02)
EPOCHS = 2000
for epoch in range(EPOCHS):
  optimizer.zero_grad()
  y_pred = x_tensor @ w
  mse = torch.mean(torch.square(y_tensor - y_pred))
  mse.backward()
  optimizer.step()
learned_w = w.detach().numpy().flatten()
print("\nKết quả sau 2000 EPOCHS")
print(f"w1 ≈ {learned_w[0]:.4f} (hệ số x^2, cần đạt: 2)")
print(f"w2 ≈ {learned_w[1]:.4f} (hệ số x,   cần đạt: -3)")
print(f"w3 ≈ {learned_w[2]:.4f} (hệ số tự do, cần đạt: 5)")

YC 1: Tạo tensor X có 24 phần tử và reshape lần lượt thành 4 x6, 2 x 12 và 2 x3 x4
Tensor x ban dau:  tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23])
Shape ban dau:  torch.Size([24])
Reshape 4 x 6: 
 tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])
Reshape 2 x 12: 
 tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])
Reshape 2 x 3 x 4: 
 tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])
YC2: CÁC PHÉP TOÁN TRÊN TENSOR A VÀ B (3 x 4)
Tensor A:
 tensor([[7., 6., 8., 5.],
        [1., 3., 8., 6.],
        [5., 3., 5., 5.]])
Tensor B:
 tensor([[9., 1., 1., 5.],
        [3., 5., 4., 5.],
        [5., 9., 2., 6.]])

Phép cộng A + B:
 tensor([[16.,  7.,  9., 10.],
 